content/blog/managing-kubernetes-infrastructure-with-dotnet-and-pulumi/index.md

```diff
@@ -0,0 +1,251 @@
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:luke-dev::basic_dotnet_kubernetes::pulumi:pulumi:Stack::basic_dotnet_kubernetes-luke-dev]
+    ~ kubernetes:core/v1:Pod: (update)
+        [id=default/pod-wyjq31t3]
+        [urn=urn:pulumi:luke-dev::basic_dotnet_kubernetes::kubernetes:core/v1:Pod::pod]
+        [provider=urn:pulumi:luke-dev::basic_dotnet_kubernetes::pulumi:providers:kubernetes::default_1_4_0_alpha_1574471668_g18eab106_dirty::127c0c24-8b12-4010-a544-5f5390f91e4e]
+      ~ metadata: {
+          ~ labels: {
+              + app: "nginx"
+            }
+        }
+
+Do you want to perform this update? yes
+Updating (luke-dev):
+
+     Type                    Name                              Status      Info
+     pulumi:pulumi:Stack     basic_dotnet_kubernetes-luke-dev              
+ ~   └─ kubernetes:core:Pod  pod                               updated     [diff: ~metadata]
+ 
+Resources:
+    ~ 1 updated
+    1 unchanged
+
+Duration: 3s
+
+Permalink: https://app.pulumi.com/lukehoban/basic_dotnet_kubernetes/luke-dev/updates/82
+```
+
+The real benefits of .NET come when we extract common code into a reusable component.  We can do that to create a new component like a `ServiceDeployment` which includes both a Kubernetes `Service` and `Deployment` using opinionated defaults.  With this, we can describe entire Kubernetes applications (100s of lines of YAML), in a short and semantically meaningful snippet of C#:
+
+```csharp
+var config = new Config();
+var isMiniKube = config.GetBoolean("isMiniKube") ?? false;
+
+var redisMaster = new ServiceDeployment("redis-master", new ServiceDeploymentArgs
+{
+    Image = "k8s.gcr.io/redis:e2e",
+    Ports = { 6379 },
+});
+
+var redisReplica = new ServiceDeployment("redis-slave", new ServiceDeploymentArgs
+{
+    Image = "gcr.io/google_samples/gb-redisslave:v1",
+    Ports = { 6379 },
+});
+
+var frontend = new ServiceDeployment("frontend", new ServiceDeploymentArgs
+{
+    Replicas = 3,
+    Image = "gcr.io/google-samples/gb-frontend:v4",
+    Ports = { 80 },
+    AllocateIPAddress = true,
+    ServiceType = isMiniKube ? "ClusterIP" : "LoadBalancer",
+});
+
+return new Dictionary<string, object?>{
+    { "frontendIp", frontend.IpAddress },
+};
+```
+
+You can check out the implementation of this `ServiceDeployment` component in the [Guestbook example](https://github.com/pulumi/examples/tree/master/kubernetes-cs-guestbook/components) in the Pulumi Examples repo on GitHub.
+
+## Cloud + Kubernetes with .NET
+
+Because Pulumi lets you work with both Kubernetes and cloud (AWS, Azure, GCP, and more), you can also create and manage the infrastructure that builds both a managed Kubernetes cluster as well as deploying applications and services into the cluster. Using a single familiar programming model, we have access to all these various cloud technologies at our fingertips.
+
+For example, we can deploy a managed Kubernetes cluster on GKE, and then deploy a Pod into it:
+
+```csharp
```

Comment:
Some of these snippets use `cs` and this one is using `csharp`. Are these equivalent?

Reply:
Yes - but I’ll make it consistent in markdown.

Reply:
`csharp` is preferred as that's what the lang chooser tabs use (on other pages). So, ideally we'd use `csharp` consistently throughout the site. But I could also tweak the chooser to support `cs` as well.